reference/JSDoc-Type-References.md

```diff
@@ -0,0 +1,242 @@
+tries to resolve type references as values if it can't find a type.
+
+Here's an example:
+
+```ts
+const FOO = "foo"
+const BAR = "bar"
+
+/** @param {FOO | BAR} type */
+function f(type) {
+}
+```
+
+In the compiler, this is resolved in two stages. Both stages use a
+JSDoc fallback. First the compiler resolves the name; then it finds
+the type for the name. Normally type resolution only looks in the type
+namespace, but when it finds nothing there, a fallback looks in the
+value namespace.
+
+In the example above, `FOO` is a value, a block-scoped `const`. During
+type resolution, there is no type named `FOO`. So instead the compiler
+uses the value named `FOO`.
+
+Now there is a problem: the value `FOO` has no type declaration, only
+a value declaration. So we now get the type of the value declaration
+instead of the type of the type declaration.
+
+I think it would help to look at an example
+that has both a value declaration and a type declaration:
+
+```ts
+var i = 0
+interface i {
+  e: 1
+  m: 1
+}
+```
+
+If you ask for the type of the value declaration of `i`, you'll get
+`number`. If you ask for the type of type declaration of `i`, you'll
+get `{ e: 1, m: 1 }`. In Javascript, you can't *write* `interface`, so
+it would just be
+
+```js
+var i = 0
+```
+
+But what if you treat `i` as a type?
+
+```js
+/** @type {i} */
+var j
+```
+
+The compiler just reuses the type of the value declaration: `number`.
+This is equivalent to inserting the `typeof` type operator to get
+`typeof i`.
+The effect is to make types more Javascripty, allowing you to specify
+objects as example types without having to learn about the difference
+between type and value:
+
+``` js
+/**
+ * @param {number} n
+ * @param {number} m
+ * @returns {-1 | 0 | 1}
+ */
+function exampleCompare(n, m) {
+}
+
+/** @param {exampleCompare} f */
+function sort(f, l) {
+  // ...
+}
+```
+
+This also works with objects, whether anonymous or an instance of a class:
+
+``` js
+const initial = {
+  frabjous: true,
+  beamish: true,
+  callooh: "callay",
+}
+
+/** @param {initial} options */
+function setup(options) {
+}
+```
+
+Relying on this fallback for complex types does produce a simpler
+type, but makes the equivalent Typescript really confusing to produce:
+
+``` js
+import { options } from './initial'
+/**
+ * @param {keyof options} k
+ * @param {options[keyof options]} v
+ */
+function demo(k, v) {
+    options[k] = v
+}
+```
+
+is equivalent to the Typescript:
+
+```ts
+import { options } from './initial'
+function demo(k: keyof typeof options, v: typeof options[keyof typeof options]) {
+}
+```
+
+Knowing where to insert the missing `typeof` operators requires more
+experience with complex type than most people can be expected to have.
+
+### Where To Find The Code ###
+
+In src/compiler/checker.ts, `getTypeFromTypeReference` is a good
+starting point. The two fallbacks are in `resolveEntityName` and
+`getTypeReferenceType`, respectively.
```

Comment:
Since the names are so confusing (not to mention the code...), maybe some note about the difference (or relationship) between `getTypeFromTypeReference` and `getTypeReferenceType`.